src/python/zquantum/core/wip/circuit/_circuit_test.py

```diff
@@ -0,0 +1,743 @@
+    gate = CustomGate(
+        matrix=sympy.Matrix(
+            [
+                [1, 0, sympy.Symbol("theta_0"), 0],
+                [0, 1, 0, sympy.Symbol("theta_1")],
+                [sympy.Symbol("theta_1"), 0, 0, 1],
+                [0, sympy.Symbol("theta_0"), 1, 0],
+            ]
+        ),
+        qubits=(0, 2),
+    )
+    circuit = Circuit(gates=[gate])
+
+    # When/Then
+    assert circuit.symbolic_params == {sympy.Symbol("theta_0"), sympy.Symbol("theta_1")}
+
+
+def test_symbolic_params_are_correct_for_multiple_gates_with_overlapping_parameters():
+    # Given
+    gate1 = CustomGate(
+        matrix=sympy.Matrix(
+            [
+                [1, 0, sympy.Symbol("theta_0"), 0],
+                [0, 1, 0, sympy.Symbol("theta_1")],
+                [sympy.Symbol("theta_1"), 0, 0, 1],
+                [0, sympy.Symbol("theta_0"), 1, 0],
+            ]
+        ),
+        qubits=(0, 2),
+    )
+    gate2 = CustomGate(
+        matrix=sympy.Matrix(
+            [
+                [sympy.Symbol("theta_0"), 0],
+                [0, sympy.Symbol("theta_1")],
+            ]
+        ),
+        qubits=(1,),
+    )
+    gate3 = CustomGate(
+        matrix=sympy.Matrix(
+            [
+                [sympy.Symbol("theta_0"), 0],
+                [0, sympy.Symbol("theta_0")],
+            ]
+        ),
+        qubits=(1,),
+    )
+    gate4 = CustomGate(
+        matrix=sympy.Matrix(
+            [
+                [sympy.Symbol("gamma_0"), 0],
+                [0, sympy.Symbol("gamma_1")],
+            ]
+        ),
+        qubits=(1,),
+    )
+
+    # When
+    circuit = Circuit(gates=[gate1, gate2, gate3, gate4])
+
+    # Then
+    assert circuit.symbolic_params == {
+        sympy.Symbol("theta_0"),
+        sympy.Symbol("theta_1"),
+        sympy.Symbol("gamma_0"),
+        sympy.Symbol("gamma_1"),
+    }
+
+
+#### __eq__ ####
+@pytest.mark.parametrize(
+    "circuit1, circuit2",
+    [
+        [
+            Circuit(gates=[]),
+            Circuit(gates=[]),
+        ],
+        [
+            Circuit(gates=[X(0), H(0), CNOT(0, 1)]),
+            Circuit(gates=[X(0), H(0), CNOT(0, 1)]),
+        ],
+        [
+            Circuit(gates=[X(0), H(0), CNOT(0, 1)]),
+            Circuit(gates=[copy.deepcopy(X(0)), H(0), CNOT(0, 1)]),
+        ],
+        [
+            Circuit(
+                gates=[
+                    X(0),
+                    H(0),
+                    CNOT(0, 1),
+                    CustomParameterizedGate,
+                ]
+            ),
+            Circuit(
+                gates=[
+                    X(0),
+                    H(0),
+                    CNOT(0, 1),
+                    CustomParameterizedGate,
+                ]
+            ),
+        ],
+    ],
+)
+def test_circuit_eq_same_gates(circuit1, circuit2):
+    """The Circuit class should be able to be able to compare two equal circuit"""
+    # When
+    are_equal = circuit1 == circuit2
+
+    # Then
+    assert are_equal
```

Comment:
Why not just `assert circuit1 == circuit2`?

Reply:
✅ d7efce0